R/backend-file.R

```diff
@@ -0,0 +1,507 @@
+                        sapply(all_items, `[[`, "user_name") %in% username)
+
+  if (is_duplicate)
+    b_file_error("cannot save secret",
+                 "The specified item is already in the keychain.")
+
+  new_item <- list(
+    service_name = service,
+    user_name = username,
+    secret = b_file_secret_encrypt(
+      password,
+      private$nonce_get(keyring),
+      private$key_get(keyring)
+    )
+  )
+
+  private$keyring_set(keyring, items = c(all_items, list(new_item)))
+  private$keyring_write_file(keyring)
+
+  invisible(self)
+}
+
+b_file_list <- function(self, private, service, keyring) {
+
+  all_items <- private$items_get(keyring)
+
+  res <- data.frame(
+    service = sapply(all_items, `[[`, "service_name"),
+    username = sapply(all_items, `[[`, "user_name"),
+    stringsAsFactors = FALSE
+  )
+
+  if (!is.null(service))
+    res[res[["service"]] == service, ]
+  else
+    res
+}
+
+b_file_keyring_create <- function(self, private, keyring)
+  private$keyring_create_direct(keyring)
+
+b_file_keyring_delete <- function(self, private, keyring) {
+
+  if (self$keyring_is_locked(keyring))
+    self$keyring_unlock(keyring)
+
+  unlink(private$keyring_file(keyring))
+
+  invisible(self)
+}
+
+b_file_keyring_lock <- function(self, private, keyring) {
+
+  assert_that(file.exists(private$keyring_file(keyring)))
+
+  private$key_unset(keyring)
+
+  invisible(self)
+}
+
+b_file_keyring_unlock <- function(self, private, keyring, password) {
+
+  private$key_set(password, keyring)
+
+  assert_that(file.exists(private$keyring_file(keyring)))
+
+  if (self$keyring_is_locked(keyring)) {
+    private$key_unset(keyring)
+    b_file_error("failed to unlock keyring")
+  }
+
+  invisible(self)
+}
+
+b_file_keyring_is_locked <- function(self, private, keyring) {
+
+  keyring <- keyring %||% private$keyring
+
+  if (!file.exists(keyring) || !private$key_is_set(keyring)) {
+    TRUE
+  } else {
+    tryCatch(
+      {
+        b_file_secret_decrypt(
+          private$check_get(keyring),
+          private$nonce_get(keyring),
+          private$key_get(keyring)
+        )
+        FALSE
+      },
+      error = function(e) {
+        if(conditionMessage(e) == "Failed to decrypt")
+          TRUE
+        else
+          signalCondition(e)
+      }
+    )
+  }
+}
+
+b_file_keyring_default <- function(self, private) {
+  private$keyring
+}
+
+b_file_keyring_set_default <- function(self, private, keyring) {
+  private$keyring <- keyring
+  invisible(self)
+}
+
+## --------------------------------------------------------------------
+## Private
+
+b_file_keyring_create_direct <- function(self, private, keyring, password,
+  nonce, items) {
+
+  file_name <- keyring %||% private$keyring
+
+  if (file.exists(file_name))
+    confirmation(paste("are you sure you want to overwrite", file_name))
+
+  private$keyring_write_file(
+    file_name,
+    nonce %||% sodium::random(24L),
```

Comment:
Instead  of  `::` , please import these functions, everywhere.

Reply:
done, for `sodium` functions and `assertthat::has_name()` (c3184679ca400b58c2f6c7a8a99a7a3c20577423); not for `yaml` functions and `utils::packageVersion()`. Do you want to import those too?